src/cache.ts

```diff
@@ -0,0 +1,114 @@
+import { KubernetesObject } from './types';
+import { Watch } from './watch';
+
+export interface ObjectCache<T> {
+    get(name: string, namespace?: string): T | null;
+    list(namespace?: string): T[];
+}
+
+export type ListCallback<T extends KubernetesObject> = (list: T[]) => void;
+
+export class ListWatch<T extends KubernetesObject> implements ObjectCache<T> {
+    private objects: T[] = [];
+    private indexCache: any = {};
+    private path: string;
+    private watch: Watch;
+    private listFn: (callback: ListCallback<T>) => void;
+
+    public constructor(path: string, watch: Watch, listFn: (callback: ListCallback<T>) => void) {
+        this.watch = watch;
+        this.listFn = listFn;
+        this.path = path;
+        this.doneHandler(null);
+    }
+
+    public get(name: string, namespace?: string): T | null {
+        let result: T | null = null;
+        for (const element of this.objects) {
+            if (element.metadata.name === name &&
+                (!namespace || element.metadata.namespace === namespace)) {
+                result = element;
```

Comment:
Can you `return element` here, or is it necessary to keep searching the list for further matching elements and return the last?

If you want to return the first that matches the criterion, consider `this.objects.find(...)`.  If you want to return the last, consider searching backward through the list, in order to avoid continued searching after finding a match.

Reply:
switched to `find` (learning more Javascript by the day :)